### PR TITLE
`MacosPopupButton` (closes #176)

### LINKED ISSUE #176: Pop-Up button 💡
A [pop-up button](https://developer.apple.com/design/human-interface-guidelines/macos/buttons/pop-up-buttons/) (often referred to as a pop-up menu) is a type of button that, when clicked, displays a menu containing a list of mutually exclusive choices.

image
image

Should be based on Flutter's `DropdownButton` widget, but with macOS-like behavior and appearance.

### PR DESCRIPTION
**Pull request by whiplashoo**
## Adds a new `MacosPopupButton` widget, trying to replicate the [Pop-Up button](https://developer.apple.com/design/human-interface-guidelines/macos/buttons/pop-up-buttons/) component from Apple's Guidelines.

**Final result:**

| Dark Theme                                 | Light Theme                                |
| ------------------------------------------ | ------------------------------------------ |
|  |  |
|  |  |
|  |  |

**Screencast:** 

https://user-images.githubusercontent.com/9117427/150102681-488c3531-791f-4de3-8613-c3cc675d97ce.mov

**Based on Flutter's Material `DropdownButton` widget, but with these modifications:**

- Different size, padding, border shadows, animations, etc., trying to match Apple design as much as possible. All these were eye-balled from pop-up buttons found in System Preferences.
- Double caret icon at the right edge of the button (done with a `CustomPainter`).
- macOS-themed styling.
- No scrollbar when menu is open.
- An up or down caret symbol for when the menu has items that cannot fit the available constraints (i.e. with a longer items list).

User can also navigate the open menu with up/down arrow keys and select an item with the Return key.

Future development can include:

- Translucency effect for the open menu.
- More tests (especially for its visual design).

## Pre-launch Checklist

- [✅] I have run `dartfmt` on all changed files 
- [✅] I have incremented the package version as appropriate and updated `CHANGELOG.md` with my changes 
- [✅] I have added/updated relevant documentation 
- [✅] I have run "optimize/organize imports" on all changed files
- [✅] I have addressed all analyzer warnings as best I could

### CODE REVIEW

**Comment by whiplashoo** [comment]
Should close #176 .

**Comment by GroovinChip** [comment]
Hello, @whiplashoo! Thank you so much for your PR, it looks amazing! I am gearing up to start working on this project again in the near future, and I see that this PR is failing some checks. Would you mind addressing that before I review, please? Cheers!

**Comment by GroovinChip** [comment]
@whiplashoo what are your thoughts on adding `MacosPopupButtonTheme` and `MacosPopupButtonThemeData` classes? This way users can set a default theme at the root level, or for specific widget trees.

**Comment by GroovinChip** [comment]
Also, does this PR contain your work from #171 ? If so then we can close that PR in favor of this one.

**Comment by whiplashoo** [comment]
@GroovinChip 

- No, it doesn't contain work from #171, they are separate branches.
- Yes, I can see how a `MacosPopupButtonTheme` could be useful. I guess it should be implemented in a similar logic like `PushButtonThemeData`, for example?

**Comment by GroovinChip** [comment]
> @whiplashoo 
> 
> * Yes, I can see how a `MacosPopupButtonTheme` could be useful. I guess it should be implemented in a similar logic like `PushButtonThemeData`, for example?

Yep, exactly

File: example/macos/Podfile.lock [hunk 1]
```diff
@@ -19,4 +19,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 6eac6b3292e5142cfc23bdeb71848a40ec51c14c
 
-COCOAPODS: 1.11.2
+COCOAPODS: 1.10.2
```
Comment (GroovinChip) [inline example/macos/Podfile.lock]:
```suggestion
COCOAPODS: 1.11.2
```

**Review by GroovinChip — CHANGES_REQUESTED** [review]
Looks great. Let's just get the Theme classes implemented and then we'll be good to go.

**Comment by whiplashoo** [comment]
In trying to implement the theme classes, I tried to use as a starting point the `PushButtonTheme` classes. 

I noticed, however, that this won't work (i.e. the color of the `PushButtons` is not yellow):

```dart
...
return MacosApp(
    title: 'macos_ui example',
    theme: MacosThemeData.light().copyWith(
      pushButtonTheme: const PushButtonThemeData( 
        color: Colors.yellow,
        disabledColor: Colors.blue,
        secondaryColor: Colors.red,
      ),
    ),
...
```

Using `PushButtonThemeData().copyWith()` doesn't help either. Isn't this the way it's expected for users to override theme colors? 

I found that in the copyWith method of `MacosThemeData`, the user-provided parameters [are not passed](https://github.com/whiplashoo/macos_ui/blob/0bf7ded8618b4d6d80d2cd07041f4124b5b3acda/lib/src/theme/macos_theme.dart#L402-L404) when constructing the `MacosThemeData`. Doing this for other `MacosThemeData` properties like `primaryColor`, `canvasColor`, etc. works properly. 

I guess this needs to be fixed first? Would be willing to tackle this on a separate issue.

**Comment by GroovinChip** [comment]
> In trying to implement the theme classes, I tried to use as a starting point the `PushButtonTheme` classes.
> 
> I noticed, however, that this won't work (i.e. the color of the `PushButtons` is not yellow):
> 
> ```dart
> ...
> return MacosApp(
>     title: 'macos_ui example',
>     theme: MacosThemeData.light().copyWith(
>       pushButtonTheme: const PushButtonThemeData( 
>         color: Colors.yellow,
>         disabledColor: Colors.blue,
>         secondaryColor: Colors.red,
>       ),
>     ),
> ...
> ```
> 
> Using `PushButtonThemeData().copyWith()` doesn't help either. Isn't this the way it's expected for users to override theme colors?
> 
> I found that in the copyWith method of `MacosThemeData`, the user-provided parameters [are not passed](https://github.com/whiplashoo/macos_ui/blob/0bf7ded8618b4d6d80d2cd07041f4124b5b3acda/lib/src/theme/macos_theme.dart#L402-L404) when constructing the `MacosThemeData`. Doing this for other `MacosThemeData` properties like `primaryColor`, `canvasColor`, etc. works properly.
> 
> I guess this needs to be fixed first? Would be willing to tackle this on a separate issue.

Oof. Yes, this looks like a bug

**Comment by whiplashoo** [comment]
Added `MacosPopupButtonTheme` and `MacosPopupButtonThemeData` classes, so user can theme `MacosPopupButton`s from the root of the app (or per widget tree).

The customizable settings are `highlightColor`, `backgroundColor`, `popupColor`. 

For a visual example, setting these colors:

```dart
return MacosApp(
  title: 'macos_ui example',
  theme: MacosThemeData.light().copyWith(
    macosPopupButtonTheme: const MacosPopupButtonThemeData(
      highlightColor: MacosColors.appleOrange,
      backgroundColor: MacosColors.appleGreen,
      popupColor: MacosColors.systemTealColor,
    ),
  ),
```

results to this `MacosPopupButton` style:

<p>
Screenshot 2022-03-08 at 9 44 43 AM

Screenshot 2022-03-08 at 9 44 53 AM
</p>

I will also open an issue to fix the bug with the other widget theme classes.

File: lib/src/buttons/popup_button.dart [hunk 2]
```diff
@@ -0,0 +1,1614 @@
...
+  bool shouldRebuildSemantics(_UpDownCaretsPainter oldDelegate) => false;
+}
+
+/// Overrides the default style of its [MacosPopupButton] descendants.
+///
+/// See also:
+///
+///  * [MacosPopupButtonThemeData], which is used to configure this theme.
+class MacosPopupButtonTheme extends InheritedTheme {
+  /// Create a [MacosPopupButtonTheme].
```
Comment (GroovinChip) [inline lib/src/buttons/popup_button.dart]:
```suggestion
  /// Creates a [MacosPopupButtonTheme].
```

File: lib/src/buttons/popup_button.dart [hunk 3]
```diff
@@ -0,0 +1,1614 @@
...
+    return MacosPopupButtonTheme(data: data, child: child);
+  }
+
+  @override
+  bool updateShouldNotify(MacosPopupButtonTheme oldWidget) =>
+      data != oldWidget.data;
+}
+
+/// A style that overrides the default appearance of
+/// [MacosPopupButton]s when it's used with [MacosPopupButtonTheme] or with the
```
Comment (GroovinChip) [inline lib/src/buttons/popup_button.dart]:
```suggestion
/// [MacosPopupButton]s when it is used with [MacosPopupButtonTheme] or with the
```

File: lib/src/buttons/popup_button.dart [hunk 4]
```diff
@@ -0,0 +1,1614 @@
...
+  /// Creates a [MacosPopupButtonThemeData].
+  const MacosPopupButtonThemeData({
+    this.highlightColor,
+    this.backgroundColor,
+    this.popupColor,
+  });
+
+  /// The default highlight color for [MacosPopupButton].
+  ///
+  /// Sets the color of the caret icons and the color of a [MacosPopupMenuItem]s' background when the mouse hovers over it.
```
Comment (GroovinChip) [inline lib/src/buttons/popup_button.dart]:
```suggestion
  /// Sets the color of the caret icons and the color of a [MacosPopupMenuItem]'s background when the mouse hovers over it.
```

File: lib/src/buttons/popup_button.dart [hunk 5]
```diff
@@ -0,0 +1,1614 @@
...
+    Color? popupColor,
+  }) {
+    return MacosPopupButtonThemeData(
+      highlightColor: highlightColor ?? this.highlightColor,
+      backgroundColor: backgroundColor ?? this.backgroundColor,
+      popupColor: popupColor ?? this.popupColor,
+    );
+  }
+
+  /// Linearly interpolate between two [MacosPopupButtonThemeData].
```
Comment (GroovinChip) [inline lib/src/buttons/popup_button.dart]:
```suggestion
  /// Linearly interpolates between two [MacosPopupButtonThemeData].
```

**Review by GroovinChip — APPROVED** [review]
LGTM. I will commit the changes requested, deal with merge conflicts, and merge.

File: pubspec.yaml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 name: macos_ui
 description: Flutter widgets and themes implementing the current macOS design language.
-version: 0.10.1
+version: 0.11.0
```
Comment (GroovinChip) [inline pubspec.yaml]:
```suggestion
version: 0.12.0
```

**Comment by GroovinChip** [comment]
Waiting on checks to complete before merging.

**Pull request merged**